dao/src/main/scala/za/co/absa/enceladus/dao/rest/CrossHostApiCaller.scala

```diff
@@ -68,20 +78,34 @@ protected class CrossHostApiCaller private(apiBaseUrls: Vector[String], maxTryCo
 
     @tailrec
     def attempt(url: String, attemptNumber: Int, urlsTried: Int): Try[T] = {
-      val result =Try {
+      val result = Try {
         fn(url)
       }.recoverWith {
-        case e @ (_: ResourceAccessException | _: RestClientException) => Failure(DaoException("Server non-responsive", e))
+        case e @ (_: ResourceAccessException | _: RestClientException) =>
+          Failure(AutoRecoverableException("Server non-responsive", e))
       }
+
       //using match instead of recoverWith to make the function @tailrec
       result match {
-        case Failure(e: RetryableException) if attemptNumber < maxTryCount =>
+
+        case Failure(e: AutoRecoverableException) if attemptNumber < maxTryCount =>
+          logFailure(e, url, attemptNumber, None)
+          attempt(url, attemptNumber + 1, urlsTried)
+
+        case Failure(e: AutoRecoverableException) if urlsTried < baseUrlsCount =>
+          val nextUrl = nextBaseUrl()
+          logFailure(e, url, attemptNumber, Option(nextUrl))
+          attempt(nextUrl, 1, urlsTried + 1)
+
+        case Failure(e: MenasException) if retryableExceptions.contains(e) && attemptNumber < maxTryCount =>
           logFailure(e, url, attemptNumber, None)
           attempt(url, attemptNumber + 1, urlsTried)
-        case Failure(e: RetryableException) if urlsTried < baseUrlsCount =>
+
+        case Failure(e: MenasException) if retryableExceptions.contains(e) && urlsTried < baseUrlsCount =>
```

Comment:
This can be done nicer than with repeated code. `retryableExceptions` (let's rename it to `optionallyRetryableExceptions`) a set of exception **classes** can be extended by exceptions that are always retried. This can be actually a private `val` of the class (let's call it `retryableExceptions`). Then the `match` can exists with the original three `case`s.
How to organize the exception class hierarchy is up to debate then. 
* First of all, the retyping of the `ResourceAccessException` and `RestClientException`seems OK.
* It can go either to `DaoException` or the new `AutoRecoverableException`
* The exceptions can all be direct descendants of the `MenasException` (or if renamed). And the `DaoException` and eventually `AutoRecoverableException` will be always added to the internal set of `retryableExceptions` (regardless if the were in the original `optionallyRetryableExceptions`)
* Or to be cleaner the hierarchy can be:
`MenasException` (still would prefer a better name) 
->`RetryableException` -> (`DaoException`, `AutoRecoverableException`)
->`OptionallyRetriableException` -> (`UnauthorizedException`, `NotFoundException`)

The later has the advantage, that the companion object of `OptionallyRetriableException` can hold the map of HTTP status codes.